### ISSUE TITLE
Scan not working in atom plugin, error in schomd?

### ISSUE THREAD

**Issue by KarlHegbloom** [ZotPlus/zotero-better-bibtex#457]
[better-bibtex_schomd_ERROR.txt](https://github.com/ZotPlus/zotero-better-bibtex/files/160162/better-bibtex_schomd_ERROR.txt)

I used wireshark to capture the error. I'm using the latest release of Juris-M.

**Comment by KarlHegbloom** [comment]
Here's another example that fails:
<info>
    <title>JM Baby Blue Style</title>
    <id>http://juris-m.github.io/styles/jm-babyblue</id>
    <link href="http://juris-m.github.io/styles/jm-babyblue" rel="self"/>

[@galloway_jr_basic_1990]

[#citation-style]: #jm-babyblue
[#bibliography]: #start
[#bibliography]: #end

**Comment by KarlHegbloom** [comment]
[better-bibtex_schomd_baby-blue_ERROR.txt](https://github.com/ZotPlus/zotero-better-bibtex/files/160188/better-bibtex_schomd_baby-blue_ERROR.txt)

Will the "skipmappings" arg fix it?

I don't know how to run my own modified copy of your plugin yet, or I'd try it before sending this.

**Comment by retorquere** [comment]
I think this error relates to differences between Zotero and Juris-M. I'll try to add Juris-M to the test setup, but I'm really strapped for time right now, so it's not going to happen immediately.

Running a modified copy is not as easy as it should be; it requires quite a few node and ruby packages to be installed to assemble the plugin, and running a modified plugin in Firefox is a non-starter ever since Mozilla started enforcing extension signing. If you do want to use a git checkout, rather than unzip as described below, please let me know and I'll outline what you need to have installed.

I think Juris-M standalone will accept unsigned plugins, in which case you should be able to use the instructions at https://www.zotero.org/support/dev/source_code under "Zotero for Firefox Git installations", except of using a git checkout you just unzip the BBT XPI, and for ID you'd use "better-bibtex@iris-advies.com".

**Comment by KarlHegbloom** [comment]
After looking more closely I think that Juris-M needs a patch also. The
method that finds the style uses the zotero.org url prefix only,  but ought
to also use their own url prefix, one first then the other as a fallback.
Maybe your code is fine? Does it work right with Zotero, but not with
Juris-M?

I'm working on developing a plugin for TeXmacs similar to the ones for Atom
or Brackets, or Openoffice. To do that I am starting by reading the code
and trying the Atom plugin,  but it doesn't work!

**Comment by retorquere** [comment]
It does work correctly for zotero. It shouldn't be too difficult to get it
to work for both, but I can't look into it myself right now.

What would you need for texmacs Integration, functionally?

**Comment by KarlHegbloom** [comment]
Well,  for some styles not requiring legal citations,  it already works
fine just using bibtex files. But I need to write legal memorandums using
bluebook style. I think that the way the Atom plugin or the Brackets plugin
works is a start. I have not yet looked over the Openoffice one. So I'll
want it to open the little dialog for picking citations,  and then I'll
want the bibliography of caselaw citations separate from one for law
journal citations,  but both must use bluebook format. I think that if I
can ask Juris-M, via Better-BibTeX (?), over an http link, to output pretty
much what it does when I use the right-click menu for "create bibliography
from item", it will require only a small amount of munging to be right.
Periods after abbreviations vs ones at end of sentence are the main fix-ups
I always have to do when I paste them in manually.

**Comment by KarlHegbloom** [comment]
PS. It would be useful to have it create a sub-collection for each
document,  for documentation,  perhaps using something like the bibtex aux
file parsing?

**Comment by retorquere** [comment]
The picker should be easy enough to do, if you can tell me the format texmacs expects.

I don't follow what you want to do with the bibliography though. From what I understand from a cursory glance through the texmacs docs it will generate the bibliography by itself. Do you want to instead have Juris-M deliver a fully-formatted bibliography and insert it instead? Does BibTeX still come into play in some way or another? If you just want an HTTP interface to some Juris-M code, you don't need BBT for that; the generic pattern for adding a new HTTP interface is simply:

```
Zotero.Server.Endpoints['/myextension/helloworld'] = function() {};
Zotero.Server.Endpoints['/myextension/helloworld'].prototype = {
  supportedMethods: ['GET'],
  init: function(url, data, sendResponseCallback) {
    /*
      paramaters are available as url.query.foo, url.query.bar etc
     */
    return sendResponseCallback(200, 'application/text', 'hello world');
  }
};
```

and you don't need all the complexity of BBT unless you want something to do with BibTeX or citation keys.

The init method can do whatever you want, including creating new collections. But how would Juris-M know what document it is being called for, unless you pass that as a parameter on the URL?

The CAYW picker likewise doesn't require BBT, although it does require some knowledge of the Zotero/Juris-M internals, as it doesn't look like it expects to be called the way I do it, so I mock a few things to get it to work. The BBT code that triggers the picker and delivers the results over HTTP is fairly easy to extend, even from another extension should you want to go that way.

**Comment by KarlHegbloom** [comment]
TeXmacs can make it's own bibliography, but only according to a set of predefined styles that are part of it, or via bibtex. I don't think it can do anything with biblatex bbl files, but it can load bibtex produced bbl files and use those; that's how it's standard way of making a bibliography works internally. The problem is that none of the bibtex styles support legal citation styles, like bluebook. For that I need the citeproc / zotero to do the formatting. So yes, I want it to give me back the fully formatted bibliography.

For some reason, it will let me run a git checkout of zotero, but not of the better-bibtex plugin. I'll look at how to build and use the standalone.

**Comment by retorquere** [comment]
In what format does texmacs expect that bibliography to be? It doesn't look like it deals with LaTeX, looks like some html-ish thing.

A git checkout of BBT will not work. It's written in a mixture of technologies (ruby, pegjs, coffeescript, some external JavaScript libraries downloaded as part of the build) to produce the final plugin, and running from a git checkout requires that git repo to be a ready-to-zip, pure javascript layout. Instead, I'd recommend to just download the latest release, unzip it (XPI files are just zipfiles), and use that directory as you would the git checkout.

If you really do want to build BBT, you'll need to install ruby (at least 2.2) and node.js. When you've done that, you can go into the BBT repo directory, and issue:

```
npm install
gem install bundler
bundle install
rake
```

and you should be good to go. You'll get an error about being unable to sign the plugin but you can ignore that.

**Comment by KarlHegbloom** [comment]
I got it to work from git checkout using your instructions and have fixed
the errors. I'll try and send a pull request after I learn more and make
sure it's right. I had to use about:config to allow unsigned plugins, then
made a file with the location of the checked out git clone named the same
as your xpi file sans extension, moved the xpi out of the extensions
directory. I have made a new method for tmbibliography that uses html
output type. Now I will work on the texmacs side of it.

**Comment by retorquere** [comment]
Super, good to know the from-scratch instructions are OK. I just checked in a minor change to a helper module; if you execute

```
bundle update
```

you can then run

```
TRAVIS_PULL_REQUEST=true rake
```
 instead of `rake` to avoid the signing error. Alternately, you can set `TRAVIS_PULL_REQUEST=true` in your environment and then run rake. This change should also cause any pull requests to build in Travis (they didn't before), so you'll have some feedback on what's happening.

**Comment by retorquere** [comment]
Travis has just started to get picky about a few gems that weren't to important anyway, so I removed them from the setup. If you want the pull requests to build, you'll need to merge from master.

**Comment by KarlHegbloom** [comment]
I've started working on a script that I want to be a drop-in replacement
for the "bibtex" command, for use by TeXmacs. The idea is to have a .bib
file exported by BetterBibTeX, to get citekeys from while writing, and then
when the bibliography generation happens, it will parse the .aux file, then
do a jsonrpc call to the schomd 'tmbibliography' method (same as
bibliography, but output is, presently, html), take the result, and
translate the html into a .bbl file. The problem is that the result loses
the mapping between the citekey and the bibliography item, which I need for
the .bbl file's \bibitem{citekey} macro. That has to be there so that
TeXmacs can make a hyperlink between the cite and the bibliography entry.

I think that ideally, BetterBibTeX ought to have a bbl output type, coming
from citeproc-js. I wonder if it's possible for citeproc-js to print that
citekey information along with each item, in some way? At the least, I need
to know the right sort-order for each list: the citekeys and the
bibliography entries, so I can map them one-to-one.

This can become a general tool then, allowing any CSL style to be used with
bibtex. (TeXmacs does not support BibLaTeX, so that's not an option.)

I think that if it can print the HTML with a class or id identifier
#citekey, it would be easy to parse from the HTML. It might also be useful
for dynamic HTML highlighting or whatever.

**Comment by retorquere** [comment]
In order:

1. I can't find the tmbibliography method in either Zotero or BBT
2. It is not hard to generate new output formats for the citation processor, see https://github.com/ZotPlus/zotero-better-bibtex/blob/master/chrome/content/zotero-better-bibtex/schomd.coffee#L4, so perhaps you don't need to parse the HTML. That markdown output format also has the keys so you can see how it can be done.
3. Having a bbl output type (if by that you mean a translator) is possible in principle, but the translators don't usually have access to the citation processor. It can be arranged, though, and then you could just have it call the output format from 2. If you rather simply want to expose a specific output format via http, that's easy. 
4. Citation keys can be accessed in the output formatter, see https://github.com/ZotPlus/zotero-better-bibtex/blob/master/chrome/content/zotero-better-bibtex/schomd.coffee#L50
5. You can print the citation key as an ID by combining 4. with https://github.com/zotero/zotero/blob/4.0/chrome/content/zotero/xpcom/citeproc.js#L13145

**Comment by KarlHegbloom** [comment]
The "tmbibliography" method is one I added to it, experimentally. I'm
renaming it "bibliographyhtml" and adding a "bbl" output format. I would
like the html output format to also have the citekey available in it. I
think that would be useful to others. How can I override only one thing
from the normal html formatter? I just want the span class csl-entry to
also have it's id attribute set to the citekey; perhaps to
id="csl-bb-${citekey}"?

**Comment by retorquere** [comment]
If you  mean `<span class="csl-entry"...`, that's assembled [here](https://github.com/zotero/zotero/blob/4.0/chrome/content/zotero/xpcom/citeproc.js#L13193). You can override it with what's called a monkey-patch, which means live-patching code. There's a few samples of that in BBT, such as [this](https://github.com/ZotPlus/zotero-better-bibtex/blob/master/chrome/content/zotero-better-bibtex/zotero-better-bibtex.coffee#L858).

Monkey patching is tricky business though. The code you're patching is called by other methods that expect it to behave a certain way; you must take care to not violate that expectation, and if the code you're patching changes behavior the monkey patch will break things. The monkey-patch could look something like this (untested):

```
Zotero.CiteProc.CSL.Output.Formats.prototype.html['@bibliography/entry'] = ((original) ->
  return (state, str) ->
    entry = original.apply(@, arguments)
    try
      citekey = Zotero.BetterBibTeX.keymanager.get({itemID: state.registry.registry[@system_id].ref.id}).citekey
      entry = entry.replace('<div class', "<div id=\"csl-bb-#{citekey}\" class")
    catch
      # monkey-patch methods MUST NOT FAIL because the code calling it will implicitly expect it not to
    return entry
  )(Zotero.CiteProc.CSL.Output.Formats.prototype.html['@bibliography/entry'])
```

**Comment by KarlHegbloom** [comment]
Any idea what's causing the string "undefined" to appear in this? I've got
it creating a bbl file, but it's imposing the string "undefined" just
before page numbers for some reason.

**Comment by retorquere** [comment]
Something is returning `undefined`; if you turn that into a string (e.g. by concatenating it to a string), you get `"undefined"`. But I don't know where it happens -- could be deep inside CiteProc.

**Comment by KarlHegbloom** [comment]
I found it. In the text_filter: it needed to have

  text = '' unless text?

... as a guard.

I tried to use that monkey patch, as-is, pasted just below the other monkey
patches, but it breaks everything. The HTML output still works, but is not
modifed by it, and neither the markdown nor the bbl formatter would operate
with it in place! I'm going to leave it where I put it, but commented off,
so you can see it in the pull request I'm developing towards being able to
make for you. I don't really need html right now anyway; I just didn't know
how easy it was to add a new output formatter, so I was trying to parse the
html, and found that the citekey is not part of it. It's a lot simpler to
just have it generate the bbl file directly.

I've forked your project on github and when it's ready, I'll push the
branch I'm working in and submit a pull request for you to look over. Give
me a day or two to get it right so less time is wasted fixing bad patches,
right?

I'll also be creating a github repository for my zotero-texmacs-integration
project. I'm not ready to check anything in yet, but as soon as I am, it
will be there for other people to try.

In that project, so far, I have a guile-2 script that uses a jsonrpc call
to the schomd interface to get a .bbl formatted bibliography given an array
of citekeys and a style. From within TeXmacs, it works as a drop-in
replacement for bibtex. I can make citations, run the bibliography update,
and it inserts a jm-babyblue formatted bibliography. It has that style
hard-coded for now, because the set! routine for setting the style in the
json it sends is still a stub. That will change soon enough, so it will
ultimately work for any CSL style.

Today I'm working on a TeXmacs .ts macros package that will use jsonrpc to
(as far as I understand it so far) an extension (that I'll need to write)
of your cayw interface, to pop up the citation search dialog, and return to
TeXmacs the things it needs to show that citation inline using whatever
style is selected. So for jm-babyblue, there's the first reference, with or
without a locator, and there's the following ones, for ibid or ibid with
locator, etc. I think I'll have it retrieve them all, and cache them the
way that the (text natbib) module presently does it, separated by some sort
of record separator, or a separate record for each type for each citekey...
then hook the update bibliography routine so that it re-runs the jsonrpc
calls to catch any updates en-batch.

Later after TeXmacs is upgraded to Guile 2, perhaps a background thread can
handle this, maybe even staying connected the way the libreoffice plugin
does, accepting calls from Juris-M / Zotero / Better BibTeX to signal
updates made there. That's not easy to do now because the (web client)
module is not in Guile 1.8 and the guile-json doesn't work in it either. So
for now it's easier to just have it only update them when I tell it to,
sort of like running a LaTeX document three times.

**Comment by retorquere** [comment]
Fixed by PR #459
